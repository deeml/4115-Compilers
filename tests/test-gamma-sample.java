import java.util.*;
import java.lang.*;

class program 
{
//input

//output

private static Random __rand = new Random(System.currentTimeMillis());
private static double gamma_sample(double alpha,double beta,double lambda)
{
double gamma = 0.0;
gamma = 0.;
if (alpha <= 0. || beta <= 0.)
{
return 0 - 1.;
System.out.println("alpha and beta must be positive.");
}
else
{
}
if (alpha < 1.)
{
while (flag == 0) {
if (p > 1.)
{
if (__rand.nextFloat() <= Math.pow(gamma,alpha - 1.))
{
flag = 1;
}
else
{
}
gamma = 0 - Math.log(b - p / alpha);
}
else
{
if (__rand.nextFloat() <= Math.pow(2.718282,0 - gamma))
{
flag = 1;
}
else
{
}
gamma = Math.pow(p,1. / alpha);
}
p = b * __rand.nextFloat();
}
b = 1 + alpha * 0.3678794;
flag = 0;
int flag = 0;
double p = 0.0;
double b = 0.0;
}
else
{
if (alpha == 1.)
{
gamma = 0 - Math.log(__rand.nextFloat());
}
else
{
gamma = alpha * y;
while (__rand.nextFloat() > Math.pow(y * 1. - y,alpha - 1.)) {
y = 0 - Math.log(__rand.nextFloat());
}
y = 0 - Math.log(__rand.nextFloat());
double y = 0.0;
}
}
return beta * gamma + lambda;
}
private static double random_sample()
{
double u = 0.0;
u = __rand.nextFloat();
if (u < 0.5)
{
return Math.sqrt(0 - 0.5 * 3.141593 * Math.log(1 - Math.pow(2 * u - 1,2)));
}
else
{
return 0 - Math.sqrt(0 - 0.5 * 3.141593 * Math.log(1 - Math.pow(1 - 2 * u,2)));
}
}

public static void main(String[] args){
__rand = new Random(1234);
int i = 0;
i = 0;
double avg = 0.0;
avg = 0.;
while (i < 1000) {
i = i + 1;
avg = avg + gamma_sample(3., 3., 1.);
}
System.out.println("close to 10");
System.out.println(avg / 1000);

return ;
}
}
