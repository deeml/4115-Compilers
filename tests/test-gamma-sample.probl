input{}
output{}
fun random_sample(): float {
  float u;
  u = rand();
  if (u<0.5) {
    return sqrt(-0.5*3.141593*log(1-(2*u-1)^2));
  } else {
    return -sqrt(-0.5*3.141593*log(1-(1-2*u)^2));
  }
}

/* Return a random number drawn from a Gamma distribution with mean alpha*beta+lamba and variance alpha*beta^2. */
fun gamma_sample(float alpha, float beta, float lambda): float {
  float gamma;
  gamma=0.0;
  if (alpha <= 0.0 or beta <= 0.0) {
     print('alpha and beta must be positive.');
     return -1.0;
  }
  if (alpha < 1.0) {
    float b;
    float p;
    int flag;
    flag=0;
    b=1+alpha*0.3678794;
    while (flag == 0) {
      p=b*rand();
      if (p>1.0) {
        gamma=-log((b-p)/alpha);
        if (rand()<=gamma^(alpha-1.0)) {
          flag=1;
        }
      }
      else {
        gamma=p^(1.0/alpha);
        if (rand()<=2.718282^(-gamma)) { 
          flag=1;
        }
      }
    }
  } else {
    if (alpha == 1.0) {
      gamma = -log(rand());
    } else {
      float y;
      y = -log(rand());
      while (rand() > (y * (1.0 - y))^(alpha - 1.0)) {
        y = -log(rand());
      }
      gamma = alpha * y;
    }
  }
  return beta*gamma+lambda;
}
model
{
  srand(1234);
  int i;
  i=0;
  float avg;
  avg=0.0;
  while (i<1000) {
    avg = avg + gamma_sample(3.0, 3.0, 1.0);
    i = i+1;
  }
  print('should be close to 10');
  print(avg/1000);
}