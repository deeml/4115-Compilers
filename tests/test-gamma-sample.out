should be close to 10
10.046462580567333
